File: crates/addr-spec/0.5.3/src/ascii.c
#include "ascii.h"
#include <string.h>

// NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
size_t ascii_escape_len(char esc_chr, const char *src, size_t len,
                        const char *cntl_chrs, size_t cntl_len) {
  size_t escape_count = 0;
  const char *end = src + len;
  for (; src < end; ++src) {
    char chr = *src;
    if (chr == esc_chr || memchr(cntl_chrs, chr, cntl_len)) {
      escape_count += 1;
    }
  }
  return len + escape_count;
}

void ascii_escape(char esc_chr, const char *src, size_t len, char *dst,
                  const char *cntl_chrs, size_t cntl_len) {
  const char *end = src + len;
  for (; src < end; ++src) {
    char chr = *src;
    if (chr == esc_chr || memchr(cntl_chrs, chr, cntl_len)) {
      *dst++ = esc_chr;
    }
    *dst++ = chr;
  }
}

File: crates/addr-spec/0.5.3/src/ascii.rs
use std::os::raw::c_char;

include!(concat!(env!("OUT_DIR"), "/ascii.rs"));

pub fn escape<const N: usize>(esc_chr: char, src: &str, cntl_chrs: [u8; N]) -> String {
    let len = unsafe {
        ascii_escape_len(
            esc_chr as c_char,
            src.as_ptr() as *const c_char,
            src.len(),
            cntl_chrs.as_ptr() as *const c_char,
            N,
        )
    };
    let mut dst = Vec::with_capacity(len);
    unsafe {
        ascii_escape(
            esc_chr as c_char,
            src.as_ptr() as *const c_char,
            src.len(),
            dst.as_mut_ptr() as *mut c_char,
            cntl_chrs.as_ptr() as *const c_char,
            N,
        );
        dst.set_len(len);
        String::from_utf8_unchecked(dst)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_escape() {
        assert_eq!(escape('\\', "", [b'"']), "");
        assert_eq!(escape('\\', "abc", [b'"']), "abc");
        assert_eq!(escape('\\', "a\\b", [b'"']), "a\\\\b");
        assert_eq!(escape('\\', "a\"b", [b'"']), "a\\\"b");
        assert_eq!(escape('\\', "a\\\"b", [b'"']), "a\\\\\\\"b");
        assert_eq!(escape('\\', "😄\"😄😄", [b'"']), "😄\\\"😄😄");
    }
}
